File: code/stabilityworkspace/ngdual.py
import numpy as np
import util
import gdual


# construct a new ngdual tuple for e^logZ * <x/Z, dx>_q
def ngdual_new_x_dx(x, q):
    if isinstance(x, tuple):
        # take the first instance of x.utp as x
        utp = np.zeros(q, dtype=np.double)
        utp[0] = np.exp(x[0]) * x[1][0]
        if q > 1:
            utp[1] = 1.0

        # normalize the utp
        Z = np.max([utp[0], 1.0])
        logZ = np.log(Z)
        utp /= Z

        return logZ, utp
    elif isinstance(x, np.ndarray):
        # take the first instance of x[:] as x
        utp = np.zeros(q, dtype=np.double)
        utp[0] = x[0]
        if q > 1:
            utp[1] = 1.0

        # normalize the utp
        Z = np.max([x[0], 1.0])
        logZ = np.log(Z)
        utp /= Z

        return logZ, utp
    else:
        # construct a new utp array of length q
        utp = np.zeros(q, dtype=np.double)
        utp[0] = x
        if q > 1:
            utp[1] = 1.0

        # normalize the utp
        Z = np.max([x, 1.0])
        logZ = np.log(Z)
        utp /= Z

        return logZ, utp


def ngdual_new_c_dx(c, q):
    # construct a new utp array of length q
    utp = np.zeros(q, dtype=np.double)
    utp[0] = c

    # normalize the utp
    Z = np.max([c, 1.0])
    logZ = np.log(Z)
    utp /= Z

    return logZ, utp

# compose two ngduals as G(F)
# note: F will need to be "denormalized", may be unstable
def ngdual_compose(G, F):
    out_utp = np.zeros_like(G[1])
    q       = out_utp.shape[0]

    # unnormalize F
    F_utp = np.copy(F[1])
    F_utp *= np.exp(F[0])

    # G doesn't need to be unnormalized
    G_utp = np.copy(G[1])

    # cache first term of G and then clear first terms of F,G
    G_0_cache = G_utp[0]
    G_utp[0]  = 0
    F_utp[0]  = 0

    # Horner's method truncated to q
    out_utp[0] = G_utp[q - 1]
    for i in range(q - 2, -1, -1):
        out_utp = np.convolve(out_utp, F_utp)[:q]
        out_utp[0] += G_utp[i]

    # restore cached values
    out_utp[0] = G_0_cache

    # handle normalization
    out_Z    = np.max(out_utp)
    out_logZ = np.log(out_Z) + G[0]
    out_utp  /= out_Z

    return out_logZ, out_utp


# compose two ngduals as G(F) where |F| = 1 or 2
def ngdual_compose_affine(G, F):
    if F[1].shape[0] <= 1:
        # composition with a constant F
        return G[0], np.copy(G[1])
    else:
        out_utp = np.copy(G[1])
        q       = out_utp.shape[0]

        # unnormalize F
        F_utp = np.copy(F[1])
        F_utp *= np.exp(F[0])

        # no need for Horner's method, utp composition uses only the 2nd and higher
        # coefficients, of which F has only 1 nonzero in this case
        out_utp *= np.power(F_utp[1], np.arange(0, q))

        # handle normalization
        out_Z    = np.max(out_utp)
        out_logZ = np.log(out_Z) + G[0]
        out_utp  /= out_Z

        return out_logZ, out_utp


# compute d^k/dx^k <f, dx>_q
# note: if |F| = q, then |out| = q - k
def ngdual_deriv(F, k):
    out_utp = np.copy(F[1])
    q       = out_utp.shape[0]

    # compute the vector of falling factorial terms in log space
    factln = util.fallingfactorialln(k, np.arange(q))

    # drop the lowest order terms of both vectors
    factln  = factln[k:]
    out_utp = out_utp[k:]

    # cache negative signs before switching to logspace
    out_utp_signs = np.sign(out_utp)
    out_utp       = np.log(np.abs(out_utp))

    out_utp = out_utp + factln # mul in logspace

    # handle normalization
    out_logZ = np.max(out_utp)
    out_utp  -= out_logZ
    out_logZ += F[0]

    # restore utp to linear space
    out_utp = out_utp_signs * np.exp(out_utp)

    return out_logZ, out_utp


def ngdual_scalar_add(F, c):
    out_utp = np.copy(F[1])

    # TODO: if F_utp[0] + (c / np.exp(F[0])) > 1, then renormalization can be handled explicitly
    out_utp[0] = out_utp[0] + (c / np.exp(F[0]))

    out_Z = np.max(out_utp)
    out_logZ = np.log(out_Z) + F[0]
    out_utp /= out_Z

    return out_logZ, out_utp


# compute <f * g, dx>_q from <f, dx>_q and <g, dx>_q
def ngdual_mul(F, G):
    assert F[1].shape[0] == G[1].shape[0]

    F_utp = np.copy(F[1])
    G_utp = np.copy(G[1])
    q     = F_utp.shape[0]

    out_utp = np.convolve(F_utp, G_utp)[:q]

    # handle normalization
    out_Z    = np.max(out_utp)
    out_logZ = np.log(out_Z) + F[0] + G[0]
    out_utp  /= out_Z

    return out_logZ, out_utp


def ngdual_scalar_mul(F, c):
    # push c into logZ
    return (F[0] + np.log(c), np.copy(F[1]))


def ngdual_scalar_mul_log(F, logc):
    # push logc into logZ
    return (F[0] + logc, np.copy(F[1]))


# compute <exp(f), dx>_q from <f, dx>_q
# note: F will currently be unnormalized first
def ngdual_exp(F):
    out_utp = np.empty_like(F[1])
    q       = out_utp.shape[0]

    # unnormalize F
    F_utp = np.copy(F[1])
    F_utp *= np.exp(F[0])

    # compute the first term of exp(f)
    out_utp[0] = np.exp(F_utp[0])

    # copy the non scalar terms of F
    F_utp_tilde = np.copy(F_utp[1:])

    F_utp_tilde *= np.arange(1, q)
    for i in xrange(1, q):
        out_utp[i] = np.sum(out_utp[:i][::-1] * F_utp_tilde[:i], axis=0) / i

    # handle normalization
    out_Z    = np.max(out_utp)
    out_logZ = np.log(out_Z)
    out_utp  /= out_Z

    return out_logZ, out_utp


# compute <log(f), dx>_q from <f, dx>_q
# note: F will currently be unnormalized first
def ngdual_log(F):
    out_utp = np.empty_like(F[1])
    q       = out_utp.shape[0]

    # unnormalize F
    F_utp = np.copy(F[1])
    F_utp *= np.exp(F[0])

    # compute the first term of log(f)
    out_utp[0] = np.log(F_utp[0])

    for i in xrange(1, q):
        out_utp[i] = (F_utp[i] * i - np.sum(F_utp[1:i][::-1] * out_utp[1:i]))
        out_utp[i] /= F_utp[0]
    out_utp[1:q] /= np.arange(1, q)

    # handle normalization
    out_Z    = np.max(out_utp)
    out_logZ = np.log(out_Z)
    out_utp  /= out_Z

    return out_logZ, out_utp


# compute <f^k, dx>_q from <f, dx>_q
# note: F will currently be unnormalized first
#TODO: broken if F_utp[0] < 0, need special case
def ngdual_pow(F, k):
    # unnormalize F
    F_utp = np.copy(F[1])
    F_utp *= np.exp(F[0])

    out_utp = gdual.gdual_exp(k * gdual.gdual_log(F_utp))

    # handle normalization
    out_Z    = np.max(out_utp)
    out_logZ = np.log(out_Z)
    out_utp  /= out_Z

    return out_logZ, out_utp


#compute <1/f, dx>_q from <f, dx>_q
# note: f will currently be unnormalized first
def ngdual_reciprocal(F):
    out_utp = np.zeros_like(F[1])
    q       = out_utp.shape[0]

    # unnormalize F
    F_utp = np.copy(F[1])
    F_utp *= np.exp(F[0])

    out_utp[0] = 1. / F_utp[0]
    for i in xrange(1, q):
        out_utp[i] = 1. / F_utp[0] * (-np.sum(out_utp[:i] * F_utp[i:0:-1], axis=0))

    # handle normalization
    out_Z    = np.max(out_utp)
    out_logZ = np.log(out_Z)
    out_utp  /= out_Z

    return out_logZ, out_utp

File: R/ngdual.R
#todo:
# * eliminate "denormalization" operations wherever possible
# * parameterize q in all methods, particularly when it is possible for F, G to have different q

if(!require("matrixStats")) install.packages("matrixStats")

###################################################
# FACTORY METHODS FOR COMMON NGDUAL CONSTRUCTIONS
###################################################

#construct an ngdual for f(x) = 1
#inputs: q, the number of derivatives
ngdual.1_dx <- function(q) {
  utp <- numeric(q)
  utp[1] <- 1.0
  
  logZ <- 0.0
  
  return(list(logZ=logZ, utp=utp))
}

#construct an ngdual for a constant
#inputs: C, the constant
#        q, the number of derivatives
ngdual.C_dx <- function(C, q) {
  utp <- numeric(q)
  utp[1] <- 1.0

  logZ <- log(C)
  
  return(list(logZ=logZ, utp=utp))
}

#construct an ngdual for a variable with respect to itself
#inputs: x, the value of the variable and
#        q, the number of derivatives
ngdual.x_dx <- function(x, q) {
  # if x is an ngdual already, extract the denormalized value of x
  if(is.list(x)) {
    x <- exp(x$logZ) * x$utp[1]
  }
  
  utp <- numeric(q)
  
  if(x > 1.0) {
    utp[1] <- 1.0
    if(q > 1) {
      #\frac{d}{dx} x = 1.0
      utp[2] <- 1.0/x #division by x is normalizing
    }
    
    logZ <- log(x)
  } else {
    utp[1] <- x
    if(q > 1) {
      #\frac{d}{dx} x = 1.0
      utp[2] <- 1.0
    }
    
    logZ <- 0.0
  }
  
  return(list(logZ=logZ, utp=utp))
}

###################################################
# NGDUAL UTILITIES
###################################################

# build an ngdual from a utp
ngdual.normalize <- function(utp) {
  Z <- max(1, abs(utp))
  utp <- utp / Z
  
  return(list(logZ=log(Z), utp=utp))
}

# build an ngdual from a utp in logspace
ngdual.normalize.log <- function(logutp) {
  logZ <- max(logutp)
  logutp <- logutp - logZ
  
  return(list(logZ=logZ, utp=exp(logutp)))
}

# convert an ngdual to a utp
ngdual.denormalize <- function(ngd) {
  return(exp(ngd$logZ) * ngd$utp)
}

###################################################
# LOG SPACE UTILITIES
###################################################

# compute the Pochhammer symbol, (x)_n, in log space
lpoch <- function(x, n) {
  return(lgamma(x + n) - lgamma(x))
}

# compute the falling factorial in log space
lfallingfactorial <- function(k, i) {
  return(lpoch(i - k + 1, k))
}

###################################################
# CORE NGDUAL OPERATIONS
###################################################

# add a scalar term to F (normalization is nontrivial)
ngdual.scalar_add <- function(F, c) {
  out.utp <- F$utp
  
  # TODO: if F$utp[1] + (c / np.exp(F$logZ)) > 1, then renormalization can be handled explicitly
  out.utp[1] <- out.utp[1] + (c / exp(F$logZ))
  
  out <- ngdual.normalize(out.utp)
  
  # add normalization from F
  out$logZ <- out$logZ + F$logZ
  
  return(out)
}

# multiply F by a scalar term
ngdual.scalar_mul <- function(F, c) {
  if(c > 0) {
    # push c into logZ
    F$logZ <- F$logZ + log(c)
    return(F)
  } else if(c < 0) {
    # invert the utp, but use -c for the normalization
    F$logZ <- F$logZ + log(-c)
    F$utp  <- -1 * F$utp
    return(F)
  } else { #c = 0
    F$logZ <- 0
    F$utp  <- numeric(length(F$utp))
    return(F)
  }
}

# multiply F by a scalar term (in log space)
ngdual.scalar_mul_log <- function(F, logc) {
  F$logZ <- F$logZ + logc
  
  return(F)
}

# compute <F * g, dx>_q from <F, dx>_q and <g, dx>_q
ngdual.mul <- function(F, G) {
  q <- length(F$utp)
  
  out.utp <- convolve(F$utp, rev(G$utp), type="o")[1:q]
  
  # browser()
  
  # R's convolve has some stability issues when some entries are zero,
  # which we correct for here (when the true (F∘G)[i] is 0)
  nz.F <- tail(which(F$utp != 0), 1) # position of last non-zero coeff in F
  nz.G <- tail(which(G$utp != 0), 1) # ditto for G
  if(nz.F + nz.G <= q) {
    out.utp[(nz.F + nz.G):q] <- 0
  }
  
  # browser()
  
  # normalize out
  out <- ngdual.normalize(out.utp)
  
  # add normalization from F, G
  out$logZ = out$logZ + F$logZ + G$logZ
  
  return(out)
}

#compute <1/F, dx>_q from <F, dx>_q
# note: F will need to be "denormalized", may be unstable
ngdual.reciprocal <- function(F) {
  q       <- length(F$utp)
  out.utp <- numeric(q)
  
  # denormalize F
  F_star.utp <- ngdual.denormalize(F)
  
  # evaluate the first term
  out.utp[1] <- 1.0 / F_star.utp[1]
  
  # from algopy
  if(q > 1) {
    for(i in seq(1, q-1)) {
      out.utp[i+1] <- out.utp[1] * (-sum(rev(F_star.utp[seq(2,i+1)]) * out.utp[seq(1,i)]))
    }
  }
  
  # normalize out
  out <- ngdual.normalize(out.utp)
  
  return(out)
}

# compute <exp(F), dx>_q from <F, dx>_q
ngdual.exp <- function(F) {
  q       <- length(F$utp)
  out.utp <- numeric(q)
  
  # subtract off a correction term (similar to logsumexp stability trick)
  correction <- exp(F$logZ) * F$utp[1]
  F_prime <- ngdual.scalar_add(F, -correction)
  
  Z <- exp(F_prime$logZ)
  
  # evaluate the first term
  out.utp[1] <- exp(F_prime$utp[1]) ^ Z
  
  # subset the derivative terms of F, mul by index
  F_prime.utp_tilde <- F_prime$utp[2:q] * seq(1, q - 1)
   
  # from algopy
  if(q > 1) {
    for(i in seq(1, q-1)) {
      out.utp[i+1] = Z * sum(rev(out.utp[seq(1,i)]) * F_prime.utp_tilde[seq(1,i)]) / i
    }
  }
  
  # normalize out
  out <- ngdual.normalize(out.utp)
  
  # restore correction
  out$logZ <- out$logZ + correction
  
  return(out)
}



# # compute <exp(F), dx>_q from <F, dx>_q but improve stability by computing each term in logspace
# ngdual.exp.logspace <- function(F) {
#   q           <- length(F$utp)
#   log.out.utp <- numeric(q)
#   
#   # compute the first term of exp(f)
#   Z <- exp(F$logZ)
#   log.expf <- Z * F$utp[1] # cache this value (in logspace)
#   log.out.utp[1] <- 0 # essentially translate F to 0
#   
#   # construct \tilde{F}, which is essentially the derivative terms of F
#   F.utp.tilde <- F$utp[2:q] * seq(1, q - 1)
#   
#   for(i in seq(2,q)) {
#     # preslice the vectors for this iteration
#     F.iter <- F.utp.tilde[seq(i-1, 1, -1)]
#     log.out.utp.iter <- log.out.utp[seq(1,i-1)]
#     
#     # remove any entries where F is zero (will become zero in logsumexp anyways)
#     log.out.utp.iter <- log.out.utp.iter[F.iter != 0]
#     F.iter           <- F.iter[F.iter != 0]
#     
#     log.out.utp[i] <- F$logZ - log(i-1) + matrixStats::logSumExp(log.out.utp.iter + log(F.iter))
#   }
#   
#   # handle normalization (explicitly)
#   out.logZ <- max(0, log.out.utp)
#   log.out.utp <- log.out.utp - out.logZ
#   out.logZ <- out.logZ + log.expf
#   
#   # convert utp out of logspace and return
#   return(list(logZ = out.logZ, utp = exp(log.out.utp)))
# }
# # ngdual.exp <- ngdual.exp.logspace

# logsumexp <- function(X, X.sign = NA) {
#   # w/o sign argument or with all nonnegative signs, default to standard logSumExp
#   # would be nice to remove this dependence
#   if(is.na(X.sign) || all(X.sign >= 0)) {
#     result.mag <- matrixStats::logSumExp(X)
#     if(result.mag == -Inf) {
#       result.sign <- 0
#     } else {
#       result.sign <- 1
#     }
#     return(c(result.mag, result.sign))
#   } else {
#     X.mag  <- abs(X)
# 
#     Z.pos <- matrixStats::logSumExp(X.mag[X.sign == 1])
#     Z.neg <- matrixStats::logSumExp(X.mag[X.sign == -1])
# 
#     if(Z.pos >= Z.neg) {
#       result.sign <- 1
#       result.mag  <- log(1 - exp(Z.neg - Z.pos)) + Z.pos
#     } else {
#       result.sign <- -1
#       result.mag  <- log(1 - exp(Z.pos - Z.neg)) + Z.neg
#     }
# 
#     return(c(result.mag, result.sign))
#   }
# }

## ported from scipy.special.logsumexp
logsumexp <- function(a, b = NA) {
  a.max <- max(a)
  
  if(!is.finite(a.max)) {
    a.max <- 0
  }
  
  if(all(!is.na(b))) {
    tmp <- b * exp(a - a.max)
  } else {
    tmp <- exp(a - a.max)
  }
  
  s <- sum(tmp)
  
  sgn <- sign(s)
  out <- log(abs(s))
  
  return(c(out, sgn))
}

# # compute <exp(F), dx>_q from <F, dx>_q but improve stability by computing each term in logspace
# ngdual.exp.logspace <- function(F) {
#   q            <- length(F$utp)
#   log.out.utp  <- numeric(q)
#   sign.out.utp <- numeric(q)
#   
#   # compute the first term of exp(f)
#   Z <- exp(F$logZ)
#   log.expf <- Z * F$utp[1] # cache this value (in logspace)
#   log.out.utp[1] <- log.expf # essentially translate F to 0
#   sign.out.utp[1] <- 1
#   
#   # construct \tilde{F}, which is essentially the derivative terms of F
#   F.utp.tilde <- F$utp[2:q] * seq(1, q - 1)
#   log.F.utp.tilde  <- log(abs(F.utp.tilde)) # log magnitude of \tilde{F}
#   sign.F.utp.tilde <- sign(F.utp.tilde)     # sign of \tilde{F}
#   
#   for(i in seq(2,q)) {
#     # preslice the vectors for this iteration
#     log.F.iter        <- log.F.utp.tilde[seq(i-1, 1, -1)]
#     sign.F.iter       <- sign.F.utp.tilde[seq(i-1, 1, -1)]
#     log.out.utp.iter  <- log.out.utp[seq(1, i-1)]
#     sign.out.utp.iter <- sign.out.utp[seq(1, i-1)]
#     
#     # remove any entries where F is zero (will become zero in logsumexp anyways)
#     log.out.utp.iter  <- log.out.utp.iter[sign.F.iter != 0]
#     sign.out.utp.iter <- sign.out.utp.iter[sign.F.iter != 0]
#     log.F.iter        <- log.F.iter[sign.F.iter != 0]
#     sign.F.iter       <- sign.F.iter[sign.F.iter != 0]
#     
#     browser()
#     
#     logsumexp.res <- logsumexp(log.out.utp.iter + log.F.iter, sign.out.utp.iter * sign.F.iter)
#     log.out.utp[i] <- F$logZ - log(i-1) + logsumexp.res[1]
#     sign.out.utp[i] <- logsumexp.res[2]
#     
#     browser()
#   }
#   
#   # handle normalization (explicitly)
#   out.logZ <- max(0, log.out.utp)
#   log.out.utp <- log.out.utp - out.logZ
#   # out.logZ <- out.logZ + log.expf
#   
#   # convert utp out of logspace and return
#   return(list(logZ = out.logZ, utp = sign.out.utp * exp(log.out.utp)))
# }

ngdual.exp.logspace <- function(F) {
  q <- length(F$utp)
  
  # output utp in signlogspace
  H.utp.logabs <- numeric(q)
  H.utp.sign   <- numeric(q)
  
  # Ft = \tilde{F} = F[2:q] * (1:q-1)
  Ft.utp.logabs <- log(abs(F$utp[2:q])) + log(seq(1, q-1))
  Ft.utp.sign   <- sign(F$utp)
  
  # compute the first term of the output utp
  H.utp.logabs[1] <- exp(F$logZ) * F$utp[1]
  H.utp.sign[1]   <- 1
  
  for(i in seq(1,q-1)) {
    # slice the vectors for this for loop iteration
    Ft.utp.logabs.iter <- Ft.utp.logabs[seq(1, i)]
    Ft.utp.sign.iter   <- Ft.utp.sign  [seq(1, i)]
    H.utp.logabs.iter  <- rev(H.utp.logabs[seq(1, i)])
    H.utp.sign.iter    <- rev(H.utp.sign  [seq(1, i)])
    
    # G = rev(H.iter) * Ft.iter
    G.logabs <- Ft.utp.logabs.iter + H.utp.logabs.iter
    G.sign   <- Ft.utp.sign.iter   * H.utp.sign.iter
    
    # logsumexp with sign
    res <- logsumexp(G.logabs, G.sign)
    H.utp.logabs[i+1] <- res[1] + F$logZ - log(i)
    H.utp.sign[i+1]   <- res[2]
  }
  
  out.logZ <- max(H.utp.logabs)
  log.out.utp <- H.utp.logabs - out.logZ
  
  return(list(logZ = out.logZ, utp = H.utp.sign * exp(log.out.utp)))
}

# compute <log(F), dx>_q from <F, dx>_q
ngdual.log <- function(F) {
  q       <- length(F$utp)
  out.utp <- numeric(q)
  
  # evaluate the first term
  out.utp[1] <- log(F$utp[1])
  
  # correct for normalization of F
  out.utp[1] <- out.utp[1] + F$logZ
  
  # from algopy
  if(q > 1) {
    for(i in seq(2, q)) {
      if(i == 2) {
        out.utp[i] <- (F$utp[i] * (i-1)) / F$utp[1]
      } else {
        out.utp[i] <- (F$utp[i] * (i-1) - sum(rev(F$utp[seq(2,i-1)]) * out.utp[seq(2,i-1)])) / F$utp[1]
      }
    }
    out.utp[2:q] <- out.utp[2:q] / seq(1, q-1)
  }
  
  # normalize out
  out <- ngdual.normalize(out.utp)
  
  return(out)
}

# compute <f^k, dx>_q from <f, dx>_q
# note: uses 3 ngdual operations, but should be more stable
#       f^k = exp(k * log(f))
ngdual.pow <- function(F, k) {
  return(ngdual.exp(ngdual.scalar_mul(ngdual.log(F), k)))
}

# compute <G, dx>_q from <G, dF>_q and <F, dx>_r
# intuitively computes <G(F), dx>_q
# note: F will need to be "denormalized", may be unstable
ngdual.compose <- function(G, F) {
  q       <- length(G$utp)
  out.utp <- numeric(q)
  
  # restore unnormalized F
  F_star.utp <- ngdual.denormalize(F)
  
  # cache first term of G, then clear first terms of F,G
  G.utp.1.cache <- G$utp[1]
  G$utp[1] <- 0
  F_star.utp[1] <- 0
  
  # Horner's method truncated to q
  out.utp[1] <- G$utp[q]
  for(i in seq(q - 1, 1, -1)) {
    out.utp <- convolve(out.utp, rev(F_star.utp), type="o")[1:q]
    out.utp[1] <- out.utp[1] + G$utp[i]
  }
  
  # restore cached values
  out.utp[1] <- G.utp.1.cache
  
  # normalize out
  out <- ngdual.normalize(out.utp)
  
  # add normalization from G
  out$logZ <- out$logZ + G$logZ
  
  return(out)
}

# compute <G, dx>_q from <G, dF>_q and <F, dx>_r
#  where r = 1 or 2
# intuitively computes <G(F), dx>_q
# note: F will need to be "denormalized", may be unstable
ngdual.affinecompose <- function(G, F) {
  q.F <- length(F$utp)
  if(q.F <= 1) {
    # composition with a constant or q == 1 F is basically a no-op
    return(G)
  } else {
    q       <- length(G$utp)
    out.utp <- G$utp
    
    # unnormalize F
    F_star.utp <- ngdual.denormalize(F)
    
    # no need for Horner's method, utp composition uses only the 2nd and higher
    # coefficients, of which F has only 1 nonzero in this case
    out.utp <- out.utp * (F_star.utp[2] ^ seq(0, q - 1))
    
    # normalize out
    out <- ngdual.normalize(out.utp)
    
    # add normalization from G
    out$logZ <- out$logZ + G$logZ
    
    return(out)
  }
}


# compute <d^k/dx^k F, dx>_{q-k} from <F, dx>_q
ngdual.deriv <- function(F, k) {
  q       <- length(F$utp)
  out.utp <- F$utp
  
  # compute the vector of falling factorial terms in log space
  factln <- lfallingfactorial(k, seq(k, q-1))
  
  # drop the lower order terms
  out.utp <- out.utp[seq(k+1, q)]
  
  # cache negative signs before switching to logspace
  out.utp.signs <- sign(out.utp)
  out.utp       <- log(abs(out.utp))
  
  # multiply out.utp and the factorials in logspace
  out.utp <- out.utp + factln
  
  # normalize out (in log space)
  out <- ngdual.normalize.log(out.utp)
  
  # add normalization from F
  out$logZ <- out$logZ + F$logZ
  
  # restore signs of utp terms
  out$utp <- out$utp * out.utp.signs
  
  return(out)
}

